tools/testutils.py

```diff
@@ -57,174 +92,210 @@ def hex_to_byte(hex_str):
-        log.error(f"Error {returncode} when executing {e.cmd}:\n{errmsg}\n{out}")
-    if log.getEffectiveLevel() <= logging.INFO:
-        if out:
-            log.info("########### PROCESS OUTPUT BEGIN:\n"
-                     f"{out}########### PROCESS OUTPUT END")
+    except subprocess.CalledProcessError as exception:
+        out = exception.stderr
+        returncode = exception.returncode
+        cmd = exception.cmd
+        # Rejoin the list for better readability.
+        if isinstance(cmd, list):
+            cmd = " ".join(cmd)
+        log.error("Error %s when executing \"%s\".", returncode, cmd)
+    except subprocess.TimeoutExpired as exception:
+        out = exception.stderr
+        returncode = FAILURE
+        cmd = exception.cmd
+        # Rejoin the list for better readability.
+        if isinstance(cmd, list):
+            cmd = " ".join(cmd)
+        log.error("Timed out when executing %s.", cmd)
+    finally:
+        outpipe.close()
+        errpipe.close()
     return ProcessResult(out, returncode)
 
 
-def kill_proc_group(proc):
+def kill_proc_group(proc: subprocess.Popen) -> None:
+    """Kill a process group and all processes associated with this group."""
     os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
 
 
-def check_root():
+def check_root() -> bool:
     """This function returns False if the user does not have root privileges.
     Caution: Only works on Unix systems"""
     return os.getuid() == 0
 
 
-class PathError(RuntimeError):
-    pass
-
-
-def check_if_file(input_path):
+def check_if_file(input_path_str: str) -> Optional[Path]:
     """Checks if a path is a file and converts the input
     to an absolute path"""
-    if not input_path:
-        log.error(sys.stderr, "input_path is None")
-        sys.exit(1)
-    input_path = Path(input_path)
+    if not input_path_str:
+        log.error("input_path is None")
+        return None
+    input_path = Path(input_path_str)
     if not input_path.exists():
-        log.error(sys.stderr, f"{input_path} does not exist")
-        sys.exit(1)
+        log.error("%s does not exist", input_path)
+        return None
     if not input_path.is_file():
-        log.error(sys.stderr, f"{input_path} is not a file")
-        sys.exit(1)
+        log.error("%s is not a file", input_path)
+        return None
     return Path(input_path.absolute())
 
 
-def check_if_dir(input_path):
+def check_if_dir(input_path_str: str) -> Optional[Path]:
     """Checks if a path is an actual directory and converts the input
     to an absolute path"""
-    if not input_path:
-        log.error(sys.stderr, "input_path is None")
-        sys.exit(1)
-    input_path = Path(input_path)
+    if not input_path_str:
+        log.error("input_path is None")
+        return None
+    input_path = Path(input_path_str)
     if not input_path.exists():
-        log.error(sys.stderr, f"{input_path} does not exist")
-        sys.exit(1)
+        log.error("%s does not exist", input_path)
```

Comment:
you can't make up your mind on using f strings of %?

Reply:
Yes, actually. It is somewhat of a contentious topic https://stackoverflow.com/questions/34619790/pylint-message-logging-format-interpolation
I think using the lazy logging interpolation with `,` is best practice, even if it uses the old formatting style. For all the other formatting, f-strings should be used.